folding-schemes/examples/external_inputs.rs

```diff
@@ -0,0 +1,213 @@
+/// together with an external input (w_i).
+///
+///        w_1     w_2     w_3     w_4     
+///        │       │       │       │      
+///        ▼       ▼       ▼       ▼      
+///       ┌─┐     ┌─┐     ┌─┐     ┌─┐     
+/// ─────►│F├────►│F├────►│F├────►│F├────►
+///  z_1  └─┘ z_2 └─┘ z_3 └─┘ z_4 └─┘ z_5
+///
+///
+/// where each F is:
+///    w_i                                        
+///     │     ┌────────────────────┐              
+///     │     │FCircuit            │              
+///     │     │                    │              
+///     └────►│ h =Hash(z_i[0],w_i)│              
+///           │ │ =Hash(v, w_i)    │              
+///  ────────►│ │                  ├───────►      
+/// z_i=[v,0] │ └──►z_{i+1}=[h, 0] │ z_{i+1}=[h,0]
+///           │                    │              
+///           └────────────────────┘
+///
+/// where each w_i value is set at the external_inputs array.
+///
+/// The last state z_i is used together with the external input w_i as inputs to compute the new
+/// state z_{i+1}.
+/// The function F will output the new state in an array of two elements, where the second element
+/// is a 0. In other words, z_{i+1} = [F([z_i, w_i]), 0], and the 0 will be replaced by w_{i+1} in
+/// the next iteration, so z_{i+2} = [F([z_{i+1}, w_{i+1}]), 0].
+#[derive(Clone, Debug)]
+pub struct ExternalInputsCircuits<F: PrimeField>
+where
+    F: Absorb,
+{
+    _f: PhantomData<F>,
+    poseidon_config: PoseidonConfig<F>,
+    external_inputs: Vec<F>,
+}
+impl<F: PrimeField> FCircuit<F> for ExternalInputsCircuits<F>
+where
+    F: Absorb,
+{
+    type Params = (PoseidonConfig<F>, Vec<F>); // where Vec<F> contains the external inputs
+
+    fn new(params: Self::Params) -> Self {
+        Self {
+            _f: PhantomData,
+            poseidon_config: params.0,
+            external_inputs: params.1,
+        }
+    }
+    fn state_len(&self) -> usize {
+        2
+    }
+
+    /// computes the next state values in place, assigning z_{i+1} into z_i, and computing the new
+    /// z_{i+1}
+    fn step_native(&self, i: usize, z_i: Vec<F>) -> Result<Vec<F>, Error> {
+        let input: [F; 2] = [z_i[0], self.external_inputs[i]];
+        let h = CRH::<F>::evaluate(&self.poseidon_config, input).unwrap();
+        Ok(vec![h, F::zero()])
+    }
+
+    /// generates the constraints for the step of F for the given z_i
+    fn generate_step_constraints(
+        &self,
+        cs: ConstraintSystemRef<F>,
+        i: usize,
+        z_i: Vec<FpVar<F>>,
+    ) -> Result<Vec<FpVar<F>>, SynthesisError> {
+        let crh_params =
+            CRHParametersVar::<F>::new_constant(cs.clone(), self.poseidon_config.clone())?;
+        let external_inputVar =
+            FpVar::<F>::new_witness(cs.clone(), || Ok(self.external_inputs[i])).unwrap();
+        let input: [FpVar<F>; 2] = [z_i[0].clone(), external_inputVar.clone()];
+        let h = CRHGadget::<F>::evaluate(&crh_params, &input)?;
+        Ok(vec![h, FpVar::<F>::zero()])
```

Comment:
It seems like it might require some heavy lifting to avoid having to put a zero as the last element of the state. As we might have to keep things that way for now, I think it could be worth it to make an assert somewhere. Something like: 
```rust
 // in folding/nova/mod.rs, within `prove_step()`
if z_i1.last() != Some(&C1::ScalarField::zero()) {
    return Err(Error::MalformedZ_i); // `Err` value: z_i should have a zero as its last element
}
```

Reply:
Yup very clear, I see how this does not apply to other examples. Thanks for the detailed explanation! 

My rationale is that it is different from implementing the circuit itself, since this is solely about "formatting logic", independent of any in-circuit computation, solely enforcing that the output is in a digestible format for the subsequent folding circuit.

I'm not sure if I'm clear and I might be wrong, in any case this isn't even a blocking issue, rather a curious inquiry! Marking this as resolved :)